### ISSUE TITLE
Caching-code in evaluator.js does not handle hash collisions

### ISSUE THREAD

**Issue by arild-haugstad** [optimizely/nuclear-js#86]
evaluate, __hasStaleValue, __cacheValue, and __isCached all seems to be built on the assumption that two objects with the same hashCode()-result must be identical.
While making this assumptions obviously provides some performance-benefits, it may on occasion lead to completely wrong results.

For the getters, I'd propose that the result of toImmutable(getter) be stored and checked with Immutable.is(), but depending on usage patterns, doing something similar for state may have an unreasonable memory-overhead.

**Comment by jtremback** [comment]
Do we know the probability of these hash collisions?

**Comment by arild-haugstad** [comment]
They *should* be "improbable" assuming that the hash-function from Immutable.js works reasonably well, which, at a glance, appears to be the case --- but trying to measure it in terms of probability can be misleading; the data (state) in question is not random; depending on application, it may happen with alarming frequency:
A trivial way of constructing hash collisions would be suggested by reading the source code at https://github.com/facebook/immutable-js/blob/master/src/Hash.js and observing that the hash values of null, undefined and false, as well as the number 0 and the empty string, are all 0 --- and with the way hash values for collections are computed, if the only change between two states is replacing one of those values with another, then the hash value stays the same.  Some applications might use null/undefined to indicate that some value is missing/unspecified --- if false or zero are valid values, then the collision on this can become a huge problem; something that happens *all the time*...

... but other than that, even intentionally constructing collisions seems rather difficult...
(Personally, I'll be using NuclearJS independently of whether this issue is "solved" --- collisions apart from the issue wrt. the different "falsy" values are too unlikely to worry about; and for those, I'll normalise to false, zero or the empty string anyway...)

**Comment by rattrayalex** [comment]
Perhaps worth adding to the docs that `evaluate`, `__hasStaleValue`, `__cacheValue`, and `__isCached` won't update if one kind of falsy value replaces another?

**Comment by HankMcCoy** [comment]
I just got bit by this issue due to the way Immutable.js handles hashing non-integer numbers prior to v3.5. In Immutable.js v3.4 (which version 1.0.5 of Nuclear JS uses) all numbers are converted to integers before calculating the hash. So `Immutable.fromJS({ foo: 0.1 })` and `Immutable.fromJS({ foo: 0.2 })` hash to the same thing. Until the next version of Nuclear JS is released using non-integer numbers in your Nuclear JS stores can result in false cache hits.

I'm getting around this for the moment by converting the value in question to 0-100 percentage, but it would definitely be nice to get a fix for this out sooner rather than later if possible.

**Comment by justanotherjones** [comment]
It's worse than just falsy values replacing each other. I unwittingly constructed a hash collision today when switching one property from `true` -> `false` and another from `false` -> `true`: http://jsfiddle.net/yx5ra4su/

Comparing all permutations of one-character keys while inverting true/false results in a ~50% collision rate: http://jsfiddle.net/wwja9ywh/ -- I also ran on all two-char keys (which takes a while) and got a similar rate.

Given that data, assuming `hashCode()` will not collide seems pretty bad.

**Comment by jordangarcia** [comment]
thanks for the fiddle @justanotherjones 

I acknowledge this is an issue and needs to be refactored.

However this has **never** surfaced at Optimizely.  Obviously the more constructed/simplified the example the higher the rate of this collision.  But not to make excuses, I will work on a fix for this as soon as possible.

There are performance implications with changing the hashing mechanism in the evaluator so whatever change must be well tested / benchmarked.

**Comment by dkozma** [comment]
I got bit by this when realizing one of the components in my app wasn't changing when it should have:

```javascript
var before = {"current":{"key1":"string","key2":0.25}};
var after = {"current":{"key1":"string","key2":0.50}};

var beforeState = Immutable.fromJS(before);
var afterState = Immutable.fromJS(after);

console.log(beforeState.hashCode(), afterState.hashCode());
// -577039243 -577039243
```

I just realized this is the same issue @HankMcCoy mentioned, however this was using Immutable 3.7.4 (the latest as of this post), after Immutable was [updated to add hashing to decimals](https://github.com/facebook/immutable-js/commit/a091a47261c36b0b251fe72d1b0ef9a0e93ffd23). So while it's not as bad as before (the hash for 0.1 now does not equal the hash for 0.2), it still can happen in later versions of Immutable.

**Comment by jordangarcia** [comment]
This will be fixed in `1.2` shortly
